cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/RecordBuilder.scala

```diff
@@ -57,19 +66,40 @@ case class RecordBuilder(namespace: String, program: String, run: String,
     // Construct a status to time map from the list of status time tuples, by keeping the earliest time of a status
     // if there exists multiple times for the same status
     val statusTimeMap = statusTimes.groupBy(_._1).map(v => (v._1, v._2.map(_._2).min))
+    // get the status with maximum time as the status
+    val status = statusTimeMap.max(Ordering[Long].on[(_,Long)](_._2))._1
     val start = statusTimeMap.get("STARTING")
     val running = statusTimeMap.get("RUNNING")
     // Get the earliest status with one of the ending statuses
     val end = statusTimeMap.filterKeys(END_STATUSES.contains).values
       .reduceOption(Math.min(_, _)) // avoid compilation error with Math.min(_, _) instead of Math.min
     val duration = end.flatMap(e => start.map(e - _))
-    val user = startInfo.map(_.user)
     val runtimeArgs = startInfo.map(_.runtimeArgs)
-    val r = Record(namespace, program, run, start, running, end, duration, user, runtimeArgs)
+    val startMethod = getStartMethod(runtimeArgs)
+    val r = Record(namespace,
+      startInfo.map(_.artifactName), startInfo.map(_.artifactVersion), startInfo.map(_.artifactScope),
+      applicationName, applicationVersion,
+      programType, program, run, status, start, running, end, duration, startInfo.map(_.user), startMethod, runtimeArgs,
+      numLogWarnings, numLogErrors, numRecordsOut)
     LOG.trace("RecordBuilder = {}", this)
     LOG.trace("Record = {}", r)
     r
   }
+
+  private def getStartMethod(runtimeArgs: Option[scala.collection.Map[String, String]]): String = {
+    if (runtimeArgs.isEmpty) return MANUAL
+    val scheduleInfoJson = runtimeArgs.get.get(SCHEDULE_INFO_KEY)
+    if (scheduleInfoJson.isEmpty) return MANUAL
+    val scheduleInfo: TriggeringScheduleInfo = GSON.fromJson(scheduleInfoJson.get, classOf[TriggeringScheduleInfo])
+    LOG.info("TriggeringScheduleInfo.name: {}", scheduleInfo.getName)
+    LOG.info("TriggeringScheduleInfo.desc: {}", scheduleInfo.getDescription)
+    val triggers = scheduleInfo.getTriggerInfos
+    if (Option(triggers).isEmpty || triggers.isEmpty) return MANUAL
+    triggers.get(0).getType match {
+      case TriggerInfo.Type.TIME => SCHEDULED
+      case _ => TRIGGERED
```

Comment:
can you use a variable such as `everythingElse` instead of `_` for default case ? will be easier to read, do we have any style guide to use `_` for scala ?

Reply:
the `_` here is not a variable, but a reserved keyword like `default` in java's switch statement

Reply:
well, i didn't say `_` is a variable. but you can use a descriptive variable there instead of `_` , since scala supports that, and asked if there is any style guide. never mind, its not critical.